rcl/include/rcl/subscription.h

```diff
@@ -0,0 +1,277 @@
+ * argument and the type associate with the subscription, via the type
+ * support, match.
+ * Passing a different type to rcl_take produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * TODO(wjwwood) blocking of take?
+ * TODO(wjwwood) pre-, during-, and post-conditions for message ownership?
+ * TODO(wjwwood) is rcl_take thread-safe?
+ * TODO(wjwwood) Should there be an rcl_message_info_t?
+ *
+ * The ros_message pointer should point to an already allocated ROS message
+ * struct of the correct type, into which the taken ROS message will be copied
+ * if one is available.
+ * If taken is false after calling, then the ROS message will be unmodified.
+ *
+ * If allocation is required when taking the message, e.g. if space needs to
+ * be allocated for a dynamically sized array in the target message, then the
+ * allocator given in the subscription options is used.
+ *
+ * The taken pointer should point to an already allocated boolean in which the
+ * result can be stored.
+ *
+ * The rmw message_info struct contains meta information about this particular
+ * message instance, like what the GUID of the publisher which published it
+ * originally or whether or not the message received from within the same
+ * process.
+ * The message_info argument should be an already allocated rmw_message_info_t
+ * structure.
+ * Passing NULL for message_info will result in the argument being ignored.
+ *
+ * \param[in] subscription the handle to the subscription from which to take
+ * \param[inout] ros_message type-erased ptr to a allocated ROS message
+ * \param[out] taken pointer to a bool, if set to false a message was not taken
+ * \param[out] message_info rmw struct which contains meta-data for the message
+ * \return RCL_RET_OK if the message was published, or
+ *         RCL_RET_INVALID_ARGUMENT if any arugments are invalid, or
+ *         RCL_RET_SUBSCRIPTION_INVALID if the subscription is invalid, or
+ *         RCL_RET_BAD_ALLOC if allocating memory failed, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+rcl_ret_t
+rcl_take(
+  const rcl_subscription_t * subscription,
+  void * ros_message,
+  bool * taken,
+  rmw_message_info_t * message_info);
+
+/// Get the topic name for the subscription.
+/* This function returns the subscription's internal topic name string.
+ * This function can fail, and therefore return NULL, if the:
+ *   - subscription is NULL
+ *   - subscription is invalid (never called init, called fini, or invalid)
+ *
+ * The returned string is only valid as long as the subscription is valid.
+ * The value of the string may change if the topic name changes, and therefore
+ * copying the string is recommended if this is a concern.
+ *
+ * This function is not thread-safe, and copying the result is not thread-safe.
+ * \TODO(wjwwood): should we provide a thread-safe copy_topic_name?
```

Comment:
The doc is clear about how long the data is valid. I would offload that to the caller.